counter/src/main/java/org/infinispan/counter/impl/weak/WeakCounterImpl.java

```diff
@@ -0,0 +1,309 @@
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.commons.api.functional.FunctionalMap;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.counter.api.CounterConfiguration;
+import org.infinispan.counter.api.CounterListener;
+import org.infinispan.counter.api.CounterType;
+import org.infinispan.counter.api.Handler;
+import org.infinispan.counter.api.WeakCounter;
+import org.infinispan.counter.exception.CounterException;
+import org.infinispan.counter.impl.entries.CounterValue;
+import org.infinispan.counter.impl.function.AddFunction;
+import org.infinispan.counter.impl.function.InitializeCounterFunction;
+import org.infinispan.counter.impl.function.ReadFunction;
+import org.infinispan.counter.impl.function.ResetFunction;
+import org.infinispan.counter.impl.listener.CounterEventImpl;
+import org.infinispan.counter.impl.listener.CounterFilterAndConvert;
+import org.infinispan.counter.impl.listener.NotificationManager;
+import org.infinispan.counter.logging.Log;
+import org.infinispan.counter.util.Utils;
+import org.infinispan.distribution.ch.ConsistentHash;
+import org.infinispan.functional.impl.FunctionalMapImpl;
+import org.infinispan.functional.impl.ReadOnlyMapImpl;
+import org.infinispan.functional.impl.ReadWriteMapImpl;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryModified;
+import org.infinispan.notifications.cachelistener.annotation.TopologyChanged;
+import org.infinispan.notifications.cachelistener.event.CacheEntryEvent;
+import org.infinispan.notifications.cachelistener.event.TopologyChangedEvent;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.util.ByteString;
+
+/**
+ * A weak consistent counter implementation.
+ * <p>
+ * Implementation: The counter is split in multiple keys and they are stored in the cache.
+ * <p>
+ * Write: A write operation will pick a key to update. If the node is a primary owner of one of the key, that key is
+ * chosen based on thread-id. This will take advantage of faster write operations. If the node is not a primary owner,
+ * one of the key in key set is chosen.
+ * <p>
+ * Read: A read operation needs to read all the key set (including the remote keys). This is slower than atomic
+ * counter.
+ * <p>
+ * Weak Read: A snapshot of all the keys values is kept locally and they are updated via cluster listeners.
+ * <p>
+ * Reset: The reset operation is <b>not</b> atomic and intermediate results may be observed.
+ *
+ * @author Pedro Ruivo
+ * @since 9.0
+ */
+@Listener(clustered = true, observation = Listener.Observation.POST, sync = false)
+public class WeakCounterImpl implements WeakCounter {
+
+   private static final Log log = LogFactory.getLog(WeakCounterImpl.class, Log.class);
+   private static final AtomicReferenceFieldUpdater<Entry, CounterValue> L1_UPDATER =
+         newUpdater(Entry.class, CounterValue.class, "snapshot");
+
+   private final Entry[] entries;
```

Comment:
I mostly agree. Although the contention will be reduced by the triangle, having multiple keys allows the node to pick one key in which it is the primary owner. This way, it allows us to pick the fastest update route.